File: frontend/src/domain/reportStatus.js
/**
 * Mirror of backend `com.example.classes.ReportStatus`.
 *
 * The forward path is UNASSIGNED -> ASSIGNED -> IN_PROGRESS -> SUBMITTED -> APPROVED -> ARCHIVED.
 * Two moves go backwards (ASSIGNED -> UNASSIGNED when the last grant is revoked, and
 * SUBMITTED -> IN_PROGRESS when a reviewer wants more work). Keep `ALLOWED_NEXT` in sync with
 * `ReportStatus.allowedNext()` so the UI never offers a move the backend will reject.
 */
export const ReportStatus = Object.freeze({
  UNASSIGNED: 'UNASSIGNED',
  ASSIGNED: 'ASSIGNED',
  IN_PROGRESS: 'IN_PROGRESS',
  SUBMITTED: 'SUBMITTED',
  APPROVED: 'APPROVED',
  ARCHIVED: 'ARCHIVED',
})

export const REPORT_STATUSES = Object.values(ReportStatus)

export const ALLOWED_NEXT = Object.freeze({
  [ReportStatus.UNASSIGNED]: [ReportStatus.ASSIGNED],
  [ReportStatus.ASSIGNED]: [ReportStatus.IN_PROGRESS, ReportStatus.UNASSIGNED],
  [ReportStatus.IN_PROGRESS]: [ReportStatus.SUBMITTED],
  [ReportStatus.SUBMITTED]: [ReportStatus.APPROVED, ReportStatus.IN_PROGRESS],
  [ReportStatus.APPROVED]: [ReportStatus.ARCHIVED],
  [ReportStatus.ARCHIVED]: [],
})

export function canMoveTo(from, to) {
  return (ALLOWED_NEXT[from] ?? []).includes(to)
}

/** Human labels for the enum values. */
export const STATUS_LABELS = Object.freeze({
  [ReportStatus.UNASSIGNED]: 'Unassigned',
  [ReportStatus.ASSIGNED]: 'Assigned',
  [ReportStatus.IN_PROGRESS]: 'In progress',
  [ReportStatus.SUBMITTED]: 'Submitted',
  [ReportStatus.APPROVED]: 'Approved',
  [ReportStatus.ARCHIVED]: 'Archived',
})

/** MUI palette key used by the status chip. Always paired with a text label, never color alone. */
export const STATUS_COLORS = Object.freeze({
  [ReportStatus.UNASSIGNED]: 'default',
  [ReportStatus.ASSIGNED]: 'info',
  [ReportStatus.IN_PROGRESS]: 'primary',
  [ReportStatus.SUBMITTED]: 'warning',
  [ReportStatus.APPROVED]: 'success',
  [ReportStatus.ARCHIVED]: 'default',
})

/** Statuses that count as "open" / current work. */
export const OPEN_STATUSES = REPORT_STATUSES.filter((s) => s !== ReportStatus.ARCHIVED)

/** Statuses that count as "completed" for the Previous Reports reference view. */
export const COMPLETED_STATUSES = [ReportStatus.APPROVED, ReportStatus.ARCHIVED]

export function isOpen(status) {
  return status !== ReportStatus.ARCHIVED
}

/** An engineer is "busy" when they hold a grant on a report in one of these states. */
export const ACTIVE_ASSIGNMENT_STATUSES = [ReportStatus.ASSIGNED, ReportStatus.IN_PROGRESS]

File: backend/springboot-service/src/main/java/com/example/classes/ReportStatus.java
package com.example.classes;

import java.util.Arrays;
import java.util.Set;

/**
 * Where a report has reached in its lifecycle.
 *
 * The database records which values are legal, through a CHECK constraint on the status column. It
 * cannot record which *moves* are legal, because a CHECK sees only the row being written and never
 * the value it replaces. That half of the rule lives here.
 *
 * The path forward is UNASSIGNED, ASSIGNED, IN_PROGRESS, SUBMITTED, APPROVED, ARCHIVED. Two moves go
 * backwards, both of them things that genuinely happen: a revoked assignment returns a report to
 * UNASSIGNED, and a reviewer may send a submitted report back for more work.
 *
 * Note that UNASSIGNED and ASSIGNED describe the same fact the report_assignment table records, from
 * the other side. The two can disagree if a grant is issued or revoked without moving the status, so
 * whichever code does one must do the other.
 */
public enum ReportStatus {

    /** Created, but nobody has been put on it yet. The state every report starts in. */
    UNASSIGNED,

    /** Somebody has been assigned, but work has not started. */
    ASSIGNED,

    /** Being worked on by its assignee. */
    IN_PROGRESS,

    /** Finished by its assignee and waiting on review. */
    SUBMITTED,

    /** Accepted by a reviewer. */
    APPROVED,

    /** Closed and kept for the record. The end of the line. */
    ARCHIVED;

    /**
     * Returns the states this one may move to.
     *
     * @return the permitted next states, empty when this state is final
     */
    public Set<ReportStatus> allowedNext() {
        return switch (this) {
            case UNASSIGNED -> Set.of(ASSIGNED);
            // Back to UNASSIGNED when the last assignment is revoked.
            case ASSIGNED -> Set.of(IN_PROGRESS, UNASSIGNED);
            case IN_PROGRESS -> Set.of(SUBMITTED);
            // Back to IN_PROGRESS when a reviewer wants more work.
            case SUBMITTED -> Set.of(APPROVED, IN_PROGRESS);
            case APPROVED -> Set.of(ARCHIVED);
            case ARCHIVED -> Set.of();
        };
    }

    /**
     * Reports whether a move to the given state is permitted from this one.
     *
     * @param next the state being moved to
     * @return true when the move is legal
     */
    public boolean canMoveTo(ReportStatus next) {
        return allowedNext().contains(next);
    }

    /**
     * Reports whether work on this report has not yet started.
     *
     * @return true for the two states that precede any work being done
     */
    public boolean isPreWork() {
        return this == UNASSIGNED || this == ASSIGNED;
    }

    /**
     * Parses a stored or submitted status value.
     *
     * Kept here rather than calling valueOf at each call site, so that an unrecognised value produces
     * a message naming the legal ones instead of a bare IllegalArgumentException.
     *
     * @param value the status name
     * @return the matching status
     * @throws IllegalArgumentException if the value is null or not a known status
     */
    public static ReportStatus of(String value) {
        return Arrays.stream(values())
            .filter(status -> status.name().equalsIgnoreCase(value))
            .findFirst()
            .orElseThrow(() -> new IllegalArgumentException(
                "Unknown status '" + value + "'; expected one of " + Arrays.toString(values())));
    }
}
